src/mplhep/plot.py

```diff
@@ -541,17 +541,25 @@ def hist2dplot(
 
     plt.sca(ax)
 
-    text_artists = []
-    if labels is not None:
+    _labels: np.ndarray | None = None
+    if (type(labels) == bool) or (np.iterable(labels)):
         if labels is True:
             _labels = H
-        elif H.shape == labels.T.shape:
-            _labels = labels.T
-        elif labels is not False:
+        elif labels is False:
+            _labels = None
+        elif H.shape == np.asarray(labels).T.shape:
+            _labels = np.asarray(labels).T
+        else:
             raise ValueError(
-                "Labels not understood, either specify a bool or a Hist-like array"
+                f"Labels input has incorrect shape  (expect: {H.shape}, got: {np.asarray(labels).T.shape})"
             )
```

Comment:
Could maybe simplify this logic a bit more with something like

```python
    _labels: np.ndarray | None = None
    if isinstance(labels, bool):
        _labels = H if labels else None
    elif np.iterable(labels):
        label_array = np.asarray(labels).T
        if H.shape == label_array.shape:
            _labels = label_array
        else:
            raise ValueError(
                f"Labels input has incorrect shape (expect: {H.shape}, got: {label_array.shape})"
            )
```

Reply:
Looks good :)